src/GeneratedHydrator/CodeGenerator/Visitor/HydratorMethodsVisitor.php

```diff
@@ -176,10 +231,67 @@ private function replaceExtract(ClassMethod $method)
                 . '"],';
         }
 
-        $body .= "\n);";
+        // None of the extract properties are optional
+        if (! array_filter($this->allowedProperties, function($conf) {
+            return $conf['extract'] === 'optional';
+        })) {
+            $body .= 'return array(';
+            foreach ($assignments as $propertyName => $a) {
+                if ($this->allowedProperties[$propertyName]['extract'] === true) {
+                    $body .= $a;
+                }
+            }
+            $body .= "\n);";
+
+            $method->stmts = $parser->parse('<?php ' . $body);
+
+            return;
+        }
+
+        // Has extract properties which are optional
+        $body .= '$ret = array(';
+        foreach ($assignments as $propertyName => $a) {
+            if ($this->allowedProperties[$propertyName]['extract'] === true) {
+                $body .= $a;
+            }
+        }
+        $body .= "\n);\n";
 
-        $method->stmts = $parser->parse('<?php ' . $body);
+        foreach ($this->accessibleProperties as $accessibleProperty) {
+            $propertyName = $accessibleProperty->getName();
+
+            if ($this->allowedProperties[$propertyName]['extract'] === 'optional') {
+                if (! $this->hasPrivatePropertiesWhichNeedExtracting || ! $accessibleProperty->isProtected()) {
+                    $propertyData = '$object->' . $propertyName;
+                } else {
+                    $propertyData = '$data["\\0*\\0' . $propertyName . '"]';
+                }
+
+                $body .= 'if (isset(' . $propertyData . ")) {\n"
```

Comment:
This is a bit doubtful if i need to check if the object property is set before i can extract it.. not sure what to do here